### ISSUE TITLE
Type inference issue related to usage of conditional type

### ISSUE THREAD

**Issue by andriy-kudrya** [Microsoft/TypeScript#25299]
**TypeScript Version:**  
2.8.1, 2.9.2, 3.0.0-dev.20180628

**Search Terms:** 
conditional type inference

**Code**

```ts
// Don't know what causes the problem so cannot come up with a shorter sample

type ActionType<P> = string & { attachPayloadTypeHack?: P & never }

/*
// irrelevant type to reproduce my issue
// but gives context that action might have no additional payload to handle

type Action<P> = P extends void
    ? { type: ActionType<P> }
    : { type: ActionType<P>, payload: P }
*/

type Handler<S, P> = P extends void
    ? (state: S) => S
    : (state: S, payload: P) => S

interface ActionHandler<S, P> {
    actionType: ActionType<P>
    handler: Handler<S, P>
}

declare function handler<S, P>(actionType: ActionType<P>, handler: Handler<S, P>): ActionHandler<S, P>

declare function createReducer<S>(
        defaultState: S,
        ...actionHandlers: ActionHandler<S /* force state S for every action handler*/, any>[]
    ): any

interface AppState {
    dummy: string
}

const defaultState: AppState = {
    dummy: ''
}

const NON_VOID_ACTION: ActionType<number> = 'NON_VOID_ACTION'
    , VOID_ACTION: ActionType<void> = 'VOID_ACTION'

handler<AppState, number>(NON_VOID_ACTION, (state, _payload) => state), // ok - state is AppState
handler<AppState, void>(VOID_ACTION, state => state) // ok - state is AppState

createReducer(
    defaultState,
    handler(NON_VOID_ACTION, (state, _payload) => state), // ok - state is AppState
    handler(VOID_ACTION, state => state) // unexpected - state is any
)
```

**Expected behavior:**
state parameter is infered to be "AppState" in the expression at the end
```ts
handler(VOID_ACTION, state => state)
```

**Actual behavior:**
state parameter is infered to be "any" 

**Playground Link:**  
[Playground](https://www.typescriptlang.org/play/index.html#src=%2F%2F%20Don't%20know%20what%20causes%20the%20problem%20so%20cannot%20come%20up%20with%20a%20shorter%20sample%0D%0A%0D%0Atype%20ActionType%3CP%3E%20%3D%20string%20%26%20%7B%20attachPayloadTypeHack%3F%3A%20P%20%26%20never%20%7D%0D%0A%0D%0A%2F*%0D%0A%2F%2F%20irrelevant%20type%20to%20reproduce%20my%20issue%0D%0A%2F%2F%20but%20gives%20context%20that%20action%20might%20have%20no%20additional%20payload%20to%20handle%0D%0A%0D%0Atype%20Action%3CP%3E%20%3D%20P%20extends%20void%0D%0A%20%20%20%20%3F%20%7B%20type%3A%20ActionType%3CP%3E%20%7D%0D%0A%20%20%20%20%3A%20%7B%20type%3A%20ActionType%3CP%3E%2C%20payload%3A%20P%20%7D%0D%0A*%2F%0D%0A%0D%0Atype%20Handler%3CS%2C%20P%3E%20%3D%20P%20extends%20void%0D%0A%20%20%20%20%3F%20(state%3A%20S)%20%3D%3E%20S%0D%0A%20%20%20%20%3A%20(state%3A%20S%2C%20payload%3A%20P)%20%3D%3E%20S%0D%0A%0D%0Ainterface%20ActionHandler%3CS%2C%20P%3E%20%7B%0D%0A%20%20%20%20actionType%3A%20ActionType%3CP%3E%0D%0A%20%20%20%20handler%3A%20Handler%3CS%2C%20P%3E%0D%0A%7D%0D%0A%0D%0Adeclare%20function%20handler%3CS%2C%20P%3E(actionType%3A%20ActionType%3CP%3E%2C%20handler%3A%20Handler%3CS%2C%20P%3E)%3A%20ActionHandler%3CS%2C%20P%3E%0D%0A%0D%0Adeclare%20function%20createReducer%3CS%3E(%0D%0A%20%20%20%20%20%20%20%20defaultState%3A%20S%2C%0D%0A%20%20%20%20%20%20%20%20...actionHandlers%3A%20ActionHandler%3CS%20%2F*%20force%20state%20S%20for%20every%20action%20handler*%2F%2C%20any%3E%5B%5D%0D%0A%20%20%20%20)%3A%20any%0D%0A%0D%0Ainterface%20AppState%20%7B%0D%0A%20%20%20%20dummy%3A%20string%0D%0A%7D%0D%0A%0D%0Aconst%20defaultState%3A%20AppState%20%3D%20%7B%0D%0A%20%20%20%20dummy%3A%20''%0D%0A%7D%0D%0A%0D%0Aconst%20NON_VOID_ACTION%3A%20ActionType%3Cnumber%3E%20%3D%20'NON_VOID_ACTION'%0D%0A%20%20%20%20%2C%20VOID_ACTION%3A%20ActionType%3Cvoid%3E%20%3D%20'VOID_ACTION'%0D%0A%0D%0Ahandler%3CAppState%2C%20number%3E(NON_VOID_ACTION%2C%20(state%2C%20_payload)%20%3D%3E%20state)%2C%20%2F%2F%20ok%20-%20state%20is%20AppState%0D%0Ahandler%3CAppState%2C%20void%3E(VOID_ACTION%2C%20state%20%3D%3E%20state)%20%2F%2F%20ok%20-%20state%20is%20AppState%0D%0A%0D%0AcreateReducer(%0D%0A%20%20%20%20defaultState%2C%0D%0A%20%20%20%20handler(NON_VOID_ACTION%2C%20(state%2C%20_payload)%20%3D%3E%20state)%2C%20%2F%2F%20ok%20-%20state%20is%20AppState%0D%0A%20%20%20%20handler(VOID_ACTION%2C%20state%20%3D%3E%20state)%20%2F%2F%20unexpected%20-%20state%20is%20any%0D%0A))
(use noImplicitAny flag)

**Related Issues:** 
Not sure #25301

**Workaround:**
When Handler is defined as follows
```ts
type Handler<S, P> = (state: S, payload: P) => S
```
it still allows to pass single argument handler and state is infered to be AppState

**Comment by weswigham** [comment]
This'd likely be fixed by #17819 if we're still interested in it. The root issue is that the apparent constraint of `P extends void
    ? (state: S) => S
    : (state: S, payload: P) => S` is `(state: S) => S | (state: S, payload : P) => S`. When looking at arrow functions to contextually type, we immediately discard signatures whose arity is smaller than what is written, which is why the two-argument case works (it narrows to one signature and is contextually typed appropriately), while the one-argument case does not (since both signatures are maybe-applicable and we give up).

**Comment by weswigham** [comment]
@RyanCavanaugh we need to revisit #17819 to fix this.
